lib/CodeGen/CGExprScalar.cpp

```diff
@@ -1570,10 +1570,21 @@ Value *ScalarExprEmitter::VisitCastExpr(CastExpr *CE) {
 
   case CK_IntToOCLSampler: {
     assert(DestTy->isSamplerT() && "CK_IntToOCLSampler cast to non sampler type");
-    return Visit(E);
+    if (!CGF.CGM.getLangOpts().CLSamplerOpaque)
+      return Visit(E);
+    if (const CastExpr* SCE = dyn_cast<CastExpr>(E)) {
```

Comment:
In some cases E is not a CastExpr, so cast will fail with assert. The same for the following 2 dyn_cast.